[13.0] Add stock_checkout_sync

guewen opened the pull request:
Move https://github.com/OCA/stock-logistics-warehouse/pull/907

Depends on:

* https://github.com/OCA/stock-logistics-warehouse/pull/1051

Regroup moves at the same location when they reach an operation type.

Example of use case:

* We use several operation types for Picking Transfers (Pick Shelves, Pick Reserve)
* We use several operation types for Packing (Pack, Pack Post)
* The "Checkout Synchronization" flag is activated on the 2 Packing operation types
* The Packing can occur in several areas (Packing/PACK1, Packing/PACK2, Packing/PACK3)

Based on these assumptions, we have these pickings, all for the same order:

* SHELF001, pick 2 goods in the shelves
* RESER001, pick 2 goods in the reserve
* PACK001, pack 3 of the goods (2 from SHELF001, 1 from RESER001)
* PPOST001, pack the remaining good of the reserve

The operator uses the button "Checkout Sync" on a pick transfer,
for instance SHELF001, and select Packing/PACK2 as destination.
We want all the lines that reach PACK001 to be updated with:

* The move lines (except the one that goes to PPOST001) must now reach
  Packing/PACK2
* The same moves' destination must reach Packing/PACK2 as well:
  we constrain the user to move the goods there as we want to move
  all the goods at the same place for the packing.

![Selection_848](https://user-images.githubusercontent.com/417223/82321125-a34bed00-99d4-11ea-84c1-7138749b6ab9.png)

Works best when used with ``stock_available_to_promise_release``:

* When using the mentioned module, we assume that we release moves (which
  creates the whole chain of moves) only when we know that we have the
  quantities in stock (otherwise the module splits the delivery). Any exception
  (which is now rare when you release this ways) that would lead to a backorder
  is handled by canceling the remaining line.
* When the module is not used, then the destination of the backorders may be
  changed at the same time

Compatible with ``stock_dynamic_routing``.

Commit:
Add stock_checkout_sync

Add an assistant to select the same destination location for all the
moves that will reach the same transfer. Generally used on packing
locations.

Commit:
Fix tests, picking type is now mandatory on routing

Commit:
Add a test ensuring new move lines have correct destination

Commit:
Remove useless check (code review)

m is supposed to be part of moves_to_update.move_dest_ids, and
moves_to_update is already a subset of self so it's always True

Commit:
Synchronize destination location on package level

Commit:
Remove stock_checkout_sync from test-requirements.txt

simahawk commented on stock_checkout_sync/models/stock_move.py:
```diff
@@ -0,0 +1,52 @@
+            for dest_picking in dest_pickings:
+                if dest_picking not in selected_pickings:
+                    selected_pickings[dest_picking] = self.env["stock.move"].browse()
+                selected_pickings[dest_picking] |= moves
```
better to use a set of ids instead?

simahawk commented on stock_checkout_sync/tests/test_checkout_sync.py:
```diff
@@ -0,0 +1,279 @@
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        cls.env = cls.env(context=dict(cls.env.context, tracking_disable=True))
```
this is done in common class

simahawk commented on stock_dynamic_routing_checkout_sync/__manifest__.py:
```diff
@@ -0,0 +1,16 @@
+# Copyright 2019 Camptocamp (https://www.camptocamp.com)
+{
+    "name": "Stock Routing Operations - Checkout Sync",
+    "summary": "Glue module",
```
this summary should be improved

Commit:
Improvements after code review

* union ids in a set instead of browse for performance in case of many
  moves
* improve summary description in glue module
* remove redundant code in test
* split test common case in "common.py": we shouldn't import a test file
  from another module

guewen commented:
@simahawk attended to your comments

simahawk commented:
@guewen reminder: remove [this line](https://github.com/OCA/wms/blob/13.0/test-requirements.txt#L2) when merged.

simahawk commented:
> @guewen reminder: remove [this line](https://github.com/OCA/wms/blob/13.0/test-requirements.txt#L2) when merged.

@guewen  could you include the change here so that we don't need another PR later?

Commit:
Remoev stock_location_children from test-requirements.txt

https://github.com/OCA/stock-logistics-warehouse/pull/855 should be
merged soon

guewen commented:
>  could you include the change here so that we don't need another PR later?

@simahawk Done, also I fixed https://github.com/OCA/stock-logistics-warehouse/pull/855 tests

simahawk commented:
/ocabot merge minor
